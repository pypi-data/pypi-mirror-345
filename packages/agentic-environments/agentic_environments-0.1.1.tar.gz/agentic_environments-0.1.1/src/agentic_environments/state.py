from typing import TypeVar


STATE = TypeVar("STATE")
