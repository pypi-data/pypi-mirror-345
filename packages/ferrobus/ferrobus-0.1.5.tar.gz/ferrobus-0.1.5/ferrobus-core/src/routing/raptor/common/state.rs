use fixedbitset::FixedBitSet;
use thiserror::Error;

use crate::{PublicTransitData, Time};

#[derive(Debug)]
pub struct RaptorState {
    // For each round and stop, we now store both the journey’s arrival time
    // and the effective boarding time (usually the trip’s departure time).
    pub arrival_times: Vec<Vec<Time>>,
    pub board_times: Vec<Vec<Time>>,
    pub marked_stops: Vec<FixedBitSet>,
    // For reporting the final journey arrival time.
    pub best_arrival: Vec<Time>,
}

#[derive(Error, Debug, PartialEq)]
pub enum RaptorError {
    #[error("Invalid stop ID")]
    InvalidStop,
    #[error("Invalid route ID")]
    InvalidRoute,
    #[error("Invalid trip index")]
    InvalidTrip,
    #[error("Invalid time value")]
    InvalidTime,
    #[error("Maximum transfers exceeded")]
    MaxTransfersExceeded,
    #[error("Invalid jorney")]
    InvalidJourney,
}

/// Result of the RAPTOR algorithm.
#[derive(Debug)]
pub enum RaptorResult {
    SingleTarget {
        arrival_time: Option<Time>,
        transfers_used: usize,
    },
    AllTargets(Vec<Time>),
}

/// Common validation and setup for RAPTOR algorithms
pub fn validate_raptor_inputs(
    data: &PublicTransitData,
    source: usize,
    target: Option<usize>,
    departure_time: Time,
) -> Result<(), RaptorError> {
    data.validate_stop(source)?;
    if let Some(target_stop) = target {
        data.validate_stop(target_stop)?;
    }
    if departure_time > 86400 * 2 {
        return Err(RaptorError::InvalidTime);
    }

    Ok(())
}

/// Get the target pruning bound for early termination
pub fn get_target_bound(state: &RaptorState, target: Option<usize>) -> Time {
    if let Some(target_stop) = target {
        state.best_arrival[target_stop]
    } else {
        Time::MAX
    }
}

impl RaptorState {
    pub fn new(num_stops: usize, max_rounds: usize) -> Self {
        RaptorState {
            arrival_times: vec![vec![Time::MAX; num_stops]; max_rounds],
            board_times: vec![vec![Time::MAX; num_stops]; max_rounds],
            marked_stops: (0..max_rounds)
                .map(|_| FixedBitSet::with_capacity(num_stops))
                .collect(),
            best_arrival: vec![Time::MAX; num_stops],
        }
    }

    pub fn update(
        &mut self,
        round: usize,
        stop: usize,
        arrival: Time,
        board: Time,
    ) -> Result<bool, RaptorError> {
        if round >= self.arrival_times.len() || stop >= self.arrival_times[0].len() {
            return Err(RaptorError::MaxTransfersExceeded);
        }
        // Only update if the new arrival time is better than what we've seen in this round
        if arrival < self.arrival_times[round][stop] {
            self.arrival_times[round][stop] = arrival;
            self.board_times[round][stop] = board;

            // Update best_arrival if this is better than any previous round
            if arrival < self.best_arrival[stop] {
                self.best_arrival[stop] = arrival;
                return Ok(true);
            }
        }
        Ok(false) // No improvement
    }
}
