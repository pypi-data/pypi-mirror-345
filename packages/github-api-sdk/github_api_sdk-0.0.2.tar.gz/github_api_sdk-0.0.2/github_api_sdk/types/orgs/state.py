

from typing_extensions import Literal, TypeAlias

__all__ = ["State"]

State: TypeAlias = Literal["open", "closed"]
