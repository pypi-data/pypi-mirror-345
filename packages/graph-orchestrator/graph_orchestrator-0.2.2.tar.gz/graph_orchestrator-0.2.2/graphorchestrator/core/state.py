from dataclasses import dataclass, field
from typing import Any, List


@dataclass
class State:
    """
    Represents the state of a process or workflow, maintaining a list of messages.

    Attributes:
        messages (List[Any]): A list to store messages related to the state.
    """

    messages: List[Any] = field(default_factory=list)

    def __repr__(self) -> str:
        """
        Provides a string representation of the State object.

        Returns:
            str: A string representation of the State, including its messages.
        """
        return f"State({self.messages})"

    def __eq__(self, other: Any) -> bool:
        """
        Checks if the current State object is equal to another object.

        Args:
            other (Any): The object to compare with.

        Returns:
            bool: True if the objects are equal, False otherwise.
        """
        if not isinstance(other, State):
            return NotImplemented
        return self.messages == other.messages
