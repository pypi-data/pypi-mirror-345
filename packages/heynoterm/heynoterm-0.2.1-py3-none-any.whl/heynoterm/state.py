from dataclasses import dataclass, field
from typing import List, Optional
from pathlib import Path
import json
from enum import Enum
import os


class Language(Enum):
    PYTHON = "python"
    MARKDOWN = "markdown"
    MATH = "math"


@dataclass
class Block:
    """Represents a single block of text displayed in the UI.

    Parameters
    ----------
    text:
        Raw text contained in the block.
    language:
        The language (or *kind*) of the block – for example *python*, *markdown*
        or *math*.  Defaults to the lower-case value of ``Language.PYTHON`` so
        that we are consistent across the code-base.  Previously we used
        ``Language.PYTHON.name`` (i.e. the **upper-case** *PYTHON*), which was
        inconsistent with the rest of the application that expects lower-case
        values coming from ``Enum.value``.  This subtle mismatch could easily
        lead to bugs when adding new features, therefore we standardise on
        ``Enum.value`` here.
    """

    text: str
    language: str = Language.PYTHON.value

    def to_terminal(self, index: int):  # noqa: ANN001 – keeps import-loop at bay
        """Convert the block to its Textual UI representation.

        The Textual components live in :pyfile:`heynoterm.block` which imports
        :pyfile:`heynoterm.state`, causing a circular dependency.  To avoid that
        we perform the import lazily inside the method.

        Parameters
        ----------
        index:
            Position of the block in the list – required so the widget can set
            a stable *id* and react correctly to focus events.
        """

        # Lazy import to prevent a circular import between *state* and *block*.
        from heynoterm.block import BlockComponent  # pylint: disable=import-outside-toplevel

        new_block = BlockComponent()
        new_block.text = self.text
        new_block.language = self.language
        new_block.index = index
        return new_block


@dataclass
class AppState:
    blocks: List[Block] = field(default_factory=list)

    def to_json(self) -> str:
        """Serialise the current state as a formatted JSON string."""

        # We cannot directly use ``dataclasses.asdict`` on the *AppState*
        # instance because that would recursively serialise *self* as well.
        # Therefore we only serialise the blocks and keep the output 100%
        # backward-compatible with the previous implementation.
        return json.dumps(
            {"blocks": [block.__dict__ for block in self.blocks]}, indent=4
        )

    @staticmethod
    def from_json(data: str) -> "AppState":
        json_data = json.loads(data)
        blocks = [Block(**block) for block in json_data.get("blocks", [])]
        return AppState(blocks=blocks)


class DataManager:
    """
    A class to manage the data of the app.
    To load, save and update the data.
    """

    def __init__(self, path: Optional[Path] = None) -> None:
        """
        Initialize the data manager.
        If `path` is provided, use that.
        Otherwise, check HEYNOTERM_STATE_PATH env var; if set, use that.
        Else default to ~/.heynoterm.json.
        """
        if path is not None:
            self.path = path
        else:
            env_path = os.getenv("HEYNOTERM_STATE_PATH")
            if env_path:
                self.path = Path(env_path)
            else:
                self.path = Path.home() / ".heynoterm.json"
        self.state: AppState = AppState()
        self.load()

    def load(self) -> None:
        """Load the data from the path."""
        if self.path.exists():
            with open(self.path, "r") as f:
                try:
                    self.state = AppState.from_json(f.read())
                except json.JSONDecodeError:
                    self.save()
        else:
            self.save()

    def save(self) -> None:
        """Save the data to the path."""
        with open(self.path, "w") as f:
            f.write(self.state.to_json())

    def add_block(
        self, block: Optional[Block] = None, index: Optional[int] = None
    ) -> None:
        """Add a block to the state."""
        if block is None:
            block = Block(text="", language=Language.MARKDOWN.value)
        if index is not None:
            self.state.blocks.insert(index, block)
        else:
            self.state.blocks.append(block)
        self.save()

    def remove_block(self, index: int) -> None:
        """Remove a block from the state."""
        self.state.blocks.pop(index)
        self.save()

    def update_block(self, index: int, block: Block) -> None:
        """Update a block in the state."""
        self.state.blocks[index] = block
        self.save()


dm = DataManager()
