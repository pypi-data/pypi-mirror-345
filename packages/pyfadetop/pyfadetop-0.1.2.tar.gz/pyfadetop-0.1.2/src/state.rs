use std::sync::{Arc, RwLock};

use anyhow::Error;
use ratatui::{
    DefaultTerminal, Frame,
    crossterm::event,
    layout::{Constraint, Direction, Layout},
    style::{Color, Style, Stylize},
    text::{Line, Span},
    widgets::{Block, Borders},
};
use tokio::sync::mpsc::Receiver;

use crate::{
    priority::SpiedRecordQueueMap,
    tabs::{
        StatefulWidgetExt,
        local_variables::{LocalVariableSelection, LocalVariableWidget},
        terminal_event::UpdateEvent,
        thread_selection::{ThreadSelectionState, ThreadSelectionWidget},
        timeline::{TimelineWidget, ViewPortBounds},
    },
};

// Add a Focus enum to track current focus
#[derive(Debug, PartialEq, Eq)]
pub enum Focus {
    ThreadList,
    Timeline,
    LogView,
}

#[derive(Debug)]
pub struct AppState {
    focus: Focus,
    thread_selection: ThreadSelectionState,
    pub(super) viewport_bound: ViewPortBounds,
    local_variable_state: LocalVariableSelection,
    pub record_queue_map: Arc<RwLock<SpiedRecordQueueMap>>,
    running: bool,
    ratio: u16,
}

impl AppState {
    fn quit(&mut self) {
        self.running = false;
    }

    pub async fn run_until_error(
        &mut self,
        mut terminal: DefaultTerminal,
        rx: &mut Receiver<UpdateEvent>,
    ) -> Result<(), Error> {
        while self.running {
            terminal.draw(|frame| self.render_full_app(frame))?;
            match rx.recv().await {
                None => {
                    break;
                }
                Some(event) => event.update_state(self)?,
            };
        }
        Ok(())
    }

    pub fn new() -> Self {
        Self {
            focus: Focus::ThreadList,
            thread_selection: Default::default(),
            record_queue_map: Default::default(),
            viewport_bound: Default::default(),
            local_variable_state: LocalVariableSelection::default(),
            running: true,
            ratio: 80,
        }
    }

    fn render_full_app(&mut self, frame: &mut Frame) {
        let out_block = {
            Block::default()
                .borders(Borders::NONE)
                .title_top(Line::from("Esc").underlined().right_aligned())
                .title_top(Line::from("Tab").underlined().left_aligned())
                .title_top(
                    Line::from(vec![
                        "Zoom ".into(),
                        Span::from("I").underlined(),
                        "n/".into(),
                        Span::from("O").underlined(),
                        "ut".into(),
                    ])
                    .left_aligned(),
                )
                .title_style(Style::default().bg(Color::Rgb(0, 0, 100)))
        };

        let inner = out_block.inner(frame.area());
        let [timeline, right] = Layout::default()
            .direction(Direction::Horizontal)
            .constraints(vec![
                Constraint::Percentage(self.ratio),
                Constraint::Percentage(100 - self.ratio),
            ])
            .areas(inner);
        let [tab_selector, locals] = Layout::default()
            .direction(Direction::Vertical)
            .constraints(vec![Constraint::Fill(1), Constraint::Fill(1)])
            .areas(right);

        match self.record_queue_map.read() {
            Ok(qmaps) => {
                self.thread_selection.update_threads(&qmaps);
                frame.render_stateful_widget(
                    ThreadSelectionWidget {
                        focused: self.focus == Focus::ThreadList,
                    }
                    .blocked(),
                    tab_selector,
                    &mut self.thread_selection,
                );
                let queue = self.thread_selection.select_thread(&qmaps);
                frame.render_stateful_widget(
                    TimelineWidget::from_queue(queue)
                        .focused(self.focus == Focus::Timeline)
                        .blocked(),
                    timeline,
                    &mut self.viewport_bound,
                );
                frame.render_stateful_widget(
                    LocalVariableWidget::from_queue(
                        queue,
                        self.viewport_bound.selected_depth as usize,
                    )
                    .focused(self.focus == Focus::LogView)
                    .blocked(),
                    locals,
                    &mut self.local_variable_state,
                );
            }
            _ => {
                self.running = false;
            }
        }

        frame.render_widget(out_block, frame.area());
    }

    pub fn handle_crossterm_events(&mut self, term_event: event::Event) -> Result<(), Error> {
        match term_event {
            event::Event::Key(key) => match (key.modifiers, key.code) {
                // Global shortcuts
                (_, event::KeyCode::Esc) => Ok(self.quit()),
                (_, event::KeyCode::Tab) => {
                    self.focus = match self.focus {
                        Focus::ThreadList => Focus::Timeline,
                        Focus::Timeline => Focus::LogView,
                        Focus::LogView => Focus::ThreadList,
                    };
                    Ok(())
                }
                (_, event::KeyCode::Char('i') | event::KeyCode::Char('o')) => {
                    Ok(self.viewport_bound.handle_zoom_event(&key))
                }
                (event::KeyModifiers::CONTROL, event::KeyCode::Right) => {
                    Ok(self.ratio = (self.ratio + 1).min(100))
                }
                (event::KeyModifiers::CONTROL, event::KeyCode::Left) => {
                    Ok(self.ratio = self.ratio.saturating_sub(1))
                }
                _ => Ok({
                    match self.focus {
                        Focus::ThreadList => self.thread_selection.handle_focused_event(&key),
                        Focus::Timeline => self.viewport_bound.handle_focused_event(&key),
                        Focus::LogView => self.local_variable_state.handle_focused_event(&key),
                    }
                }),
            },
            _ => Ok(()),
        }
    }
}
