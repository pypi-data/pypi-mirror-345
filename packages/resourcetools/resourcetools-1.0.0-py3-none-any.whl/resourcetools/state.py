from types import ModuleType

resource_package: ModuleType | None = None
