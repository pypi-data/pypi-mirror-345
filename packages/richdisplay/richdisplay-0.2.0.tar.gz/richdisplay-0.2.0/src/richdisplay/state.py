'''
State and Args context management with centralized authentication.
Provides stack-based contexts that rely on a single AuthManager to generate, verify, and revoke tokens.
'''
from dataclasses import dataclass
from threading import local
from typing import Optional, List
import secrets
import uuid


class AuthManager:
    """
    Centralized token management for all contexts.
    """
    _valid_tokens: set[str] = set()

    @classmethod
    def generate_token(cls) -> str:
        """
        Create and register a new secure token.
        """
        auth_token = secrets.token_hex(16)
        cls._valid_tokens.add(auth_token)
        return auth_token

    @classmethod
    def verify_token(cls, auth_token: str) -> bool:
        """
        Check if the token is valid.
        """
        return auth_token in cls._valid_tokens

    @classmethod
    def revoke_token(cls, auth_token: str) -> None:
        """
        Invalidate a token so it can no longer be used.
        """
        cls._valid_tokens.discard(auth_token)


@dataclass(frozen=True)
class StateSnapshot:
    """
    Immutable snapshot of the sys context.
    """
    debug: bool
    log_to_db: bool
    log_level: int
    token: str


@dataclass(frozen=True)
class ArgsSnapshot:
    """
    Immutable snapshot of the args context.
    """
    clear: bool
    debug: bool
    db: bool
    db_path: str
    log_to_db: bool
    get_logs: int
    export: str
    log_level: int
    token: str


class BaseContext:
    """
    Base for context stacks with token verification.
    """
    _local = local()
    _stack_name: str
    _snapshot_type: type

    @classmethod
    def _init_stack(cls):
        if not hasattr(cls._local, cls._stack_name):
            setattr(cls._local, cls._stack_name, [])  # type: ignore

    @classmethod
    def _current(cls):
        cls._init_stack()
        stack = getattr(cls._local, cls._stack_name)
        return stack[-1] if stack else None

    @classmethod
    def _push(cls, snapshot):
        cls._init_stack()
        getattr(cls._local, cls._stack_name).append(snapshot)

    @classmethod
    def _pop(cls, auth_token: str):
        cls._init_stack()
        stack = getattr(cls._local, cls._stack_name)
        if not stack:
            raise RuntimeError(f"No {cls.__name__} to pop.")
        current = stack[-1]
        if current.token != auth_token or not AuthManager.verify_token(auth_token):
            raise PermissionError(f"Invalid token for {cls.__name__} pop.")
        stack.pop()


class SysContext(BaseContext):
    """
    System-level context for debug and logging state.
    """
    _stack_name = uuid.uuid4().hex

    @classmethod
    def get_debug(cls) -> bool:
        state = cls._current()
        return state.debug if state else False

    @classmethod
    def get_log_to_db(cls) -> bool:
        state = cls._current()
        return state.log_to_db if state else False

    @classmethod
    def get_log_level(cls) -> int:
        state = cls._current()
        return state.log_level if state else 10

    @classmethod
    def push(
            cls, *,
            auth_token: str,
            debug: Optional[bool] = None,
            log_to_db: Optional[bool] = None,
            log_level: Optional[int] = None):
        if not AuthManager.verify_token(auth_token):
            raise PermissionError("Invalid token for SysContext push.")
        base = cls._current() or StateSnapshot(False, False, 1, auth_token)
        debug_flag = debug if debug is not None else base.debug
        level = 0 if debug_flag else (
            log_level if log_level is not None else base.log_level)
        if not isinstance(level, int) or not 10 <= level <= 50:
            level = 10
        snapshot = StateSnapshot(
            debug_flag,
            log_to_db if log_to_db is not None else base.log_to_db,
            level,
            auth_token
        )
        cls._push(snapshot)

    @classmethod
    def pop(cls, auth_token: str):
        cls._pop(auth_token)


class ArgsContext(BaseContext):
    """
    Argument parsing context for CLI flags.
    """
    _stack_name = uuid.uuid4().hex

    @classmethod
    def get(cls) -> ArgsSnapshot:
        current = cls._current()
        if current:
            return current
        return ArgsSnapshot(False, False, False, 'logs.db', False, 0, '', 10, '')

    @classmethod
    def push(cls, *, auth_token: str, clear: Optional[bool] = None,
             debug: Optional[bool] = None, db: Optional[bool] = None,
             db_path: Optional[str] = None, log_to_db: Optional[bool] = None,
             get_logs: Optional[int] = None, export: Optional[str] = None,
             log_level: Optional[int] = None):
        if not AuthManager.verify_token(auth_token):
            raise PermissionError("Invalid token for ArgsContext push.")
        base = cls._current() or ArgsSnapshot(
            False, False, False, 'logs.db', False, 0, '', 10, auth_token)
        snapshot = ArgsSnapshot(
            clear=clear if clear is not None else base.clear,
            debug=debug if debug is not None else base.debug,
            db=db if db is not None else base.db,
            db_path=db_path if db_path is not None else base.db_path,
            log_to_db=log_to_db if log_to_db is not None else base.log_to_db,
            get_logs=get_logs if get_logs is not None else base.get_logs,
            export=export if export is not None else base.export,
            log_level=log_level if log_level is not None else base.log_level,
            token=auth_token
        )
        cls._push(snapshot)

    @classmethod
    def pop(cls, auth_token: str):
        cls._pop(auth_token)


if __name__ == "__main__":
    # Centralized token handling
    token = AuthManager.generate_token()

    # SysContext demonstration
    SysContext.push(auth_token=token, debug=True)
    print(SysContext.get_debug(), SysContext.get_log_level())
    SysContext.pop(auth_token=token)

    # ArgsContext demonstration
    ArgsContext.push(auth_token=token, debug=True, get_logs=5)
    args = ArgsContext.get()
    print(args.debug, args.get_logs)
    ArgsContext.pop(token)

    # Revoke when done
    AuthManager.revoke_token(token)
