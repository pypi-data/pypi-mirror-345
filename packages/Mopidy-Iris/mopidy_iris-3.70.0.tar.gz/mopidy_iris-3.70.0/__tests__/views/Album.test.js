import React from 'react';
import { render } from '../test-wrapper';
import Album from '../../src/js/views/Album';
import { state as mockState } from '../state';

jest.mock('react-dnd', () => ({
  ...jest.requireActual('react-dnd'),
  useDrag: () => jest.fn(),
  useDrop: () => jest.fn(),
}));
jest.mock('redux-persist', () => ({
  ...jest.requireActual('redux-persist'),
  persistReducer: jest.fn().mockImplementation((config, reducers) => reducers),
}));
jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockImplementation((func) => func(mockState)),
  useDispatch: () => jest.fn(),
  connect: jest.fn(fn => fn()),
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    uri: 'bG9jYWw6YWxidW06bWQ1OjY2ZmJlYTM1OTNiYTk2YTE1YTlkNDE4OGJlYmFiNTBi',
  }),
}));

beforeEach(() => {  
  window.IntersectionObserver = jest.fn(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
  }));
});

describe('<Album />', () => {
  it('should render accurately', () => {
    const result = render(
      <Album />,
      { initialState: mockState },
    ).toJSON();
    expect(result).toMatchSnapshot();
  });
});
