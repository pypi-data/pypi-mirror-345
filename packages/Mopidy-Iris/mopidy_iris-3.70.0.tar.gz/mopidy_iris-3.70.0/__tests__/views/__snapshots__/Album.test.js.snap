// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Album /> should render accurately 1`] = `
<div
  className="view album-view content-wrapper preserve-3d"
  data-qa-file="Album"
  data-qa-node="div"
>
  <div
    className="thumbnail-wrapper"
    data-qa-file="Album"
    data-qa-node="div"
  >
    <div
      className="thumbnail thumbnail--loaded thumbnail--large"
      data-qa-file="Thumbnail"
      data-qa-node="div"
    >
      <i
        className="icon icon--material thumbnail__placeholder"
        data-qa-file="Icon"
        data-qa-node="i"
      >
        album
      </i>
      <div
        className="thumbnail__image"
        data-qa-file="Thumbnail"
        data-qa-node="div"
        style={
          {
            "backgroundImage": "url("/local/17338e740316f18dbb5e3331ac6be6c1-500x500.jpeg")",
          }
        }
      />
      <div
        className="thumbnail__image thumbnail__image--glow"
        data-qa-file="Thumbnail"
        data-qa-node="div"
        style={
          {
            "backgroundImage": "url("/local/17338e740316f18dbb5e3331ac6be6c1-500x500.jpeg")",
          }
        }
      />
      <div
        className="thumbnail__actions"
        data-qa-file="Thumbnail"
        data-qa-node="div"
      >
        <a
          className="thumbnail__actions__item thumbnail__actions__item--zoom "
          data-qa-file="Link"
          data-qa-node="RouterLink"
          href="/modal/image-zoom?url=/local/17338e740316f18dbb5e3331ac6be6c1-500x500.jpeg"
          onClick={[Function]}
        >
          <i
            className="icon icon--material "
            data-qa-file="Icon"
            data-qa-node="i"
          >
            search
          </i>
        </a>
      </div>
    </div>
  </div>
  <div
    className="title"
    data-qa-file="Album"
    data-qa-node="div"
  >
    <h1
      data-qa-file="Album"
      data-qa-node="h1"
    >
      Sirens of the Sea
    </h1>
    <ul
      className="details details--one-line"
      data-qa-file="Album"
      data-qa-node="ul"
    >
      <li
        className="source"
        data-qa-file="Album"
        data-qa-node="li"
      >
        <span
          aria-hidden={true}
          className="fa fa-folder icon icon--fontawesome source"
          data-qa-file="Icon"
          data-qa-node="FontAwesome"
          type="fontawesome"
        />
      </li>
      <li
        data-qa-file="Album"
        data-qa-node="li"
      >
        <span
          className="undefined links-sentence"
          data-qa-file="LinksSentence"
          data-qa-node="span"
        >
          <span
            data-qa-file="LinksSentence"
            data-qa-node="span"
          >
            <a
              className="links-sentence__item links-sentence__item--link "
              data-qa-file="Link"
              data-qa-node="RouterLink"
              href="/artist/bG9jYWw6YXJ0aXN0Om1kNTo0ZjZlNGY5NzllMmM0MGM1ZTZhZDE3MzU4MDRjMjliYw=="
              onClick={[Function]}
            >
              Above & Beyond
            </a>
          </span>
        </span>
      </li>
      <li
        data-qa-file="Album"
        data-qa-node="li"
      >
        2008
      </li>
      <li
        data-qa-file="Album"
        data-qa-node="li"
      >
        0 tracks
      </li>
      <li
        data-qa-file="Album"
        data-qa-node="li"
      >
        0 
        mins
      </li>
    </ul>
  </div>
  <div
    className="actions"
    data-qa-file="Album"
    data-qa-node="div"
  >
    <button
      className="button   button--primary"
      data-qa-file="Button"
      data-qa-node="Element"
      onClick={[Function]}
      type="button"
    >
      Play
    </button>
    <button
      className="button context-menu-trigger mouse-contextable touch-contextable  button--default  button--icon"
      data-qa-file="Button"
      data-qa-node="Element"
      onClick={[Function]}
      type="button"
    >
      <i
        className="icon icon--material "
        data-qa-file="Icon"
        data-qa-node="i"
      >
        more_horiz
      </i>
    </button>
  </div>
  <section
    className="list-wrapper"
    data-qa-file="Album"
    data-qa-node="section"
  >
    <h4
      className="no-bottom-margin"
      data-qa-file="Album"
      data-qa-node="h4"
    >
      Tracks
      <div
        className="actions-wrapper"
        data-qa-file="Album"
        data-qa-node="div"
      >
        <span
          className="filter-field "
          data-qa-file="FilterField"
          data-qa-node="span"
        >
          <form
            data-qa-file="FilterField"
            data-qa-node="form"
            onSubmit={[Function]}
          >
            <input
              data-qa-file="FilterField"
              data-qa-node="input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Filter"
              type="text"
              value=""
            />
            <i
              className="icon icon--material "
              data-qa-file="Icon"
              data-qa-node="i"
              onClick={null}
            >
              search
            </i>
          </form>
        </span>
        <div
          className="dropdown-field "
          data-qa-file="DropdownField"
          data-qa-node="div"
          data-uid="sort"
        >
          <div
            className="dropdown-field__label"
            data-qa-file="DropdownField"
            data-qa-node="div"
            onClick={[Function]}
          >
            <i
              className="icon icon--material "
              data-qa-file="Icon"
              data-qa-node="i"
            >
              swap_vert
            </i>
            <span
              className="text"
              data-qa-file="DropdownField"
              data-qa-node="span"
            >
              <span
                className="dropdown-field__label__value"
                data-qa-file="DropdownField"
                data-qa-node="span"
              >
                Track number
              </span>
              <span
                className="dropdown-field__label__name"
                data-qa-file="DropdownField"
                data-qa-node="span"
              >
                Sort
              </span>
            </span>
          </div>
          <div
            className="dropdown-field__options"
            data-qa-file="DropdownField"
            data-qa-node="div"
          >
            <div
              className="dropdown-field__options__liner"
              data-qa-file="DropdownField"
              data-qa-node="div"
            >
              <div
                className="dropdown-field__options__item "
                data-qa-file="DropdownField"
                data-qa-node="div"
                onClick={[Function]}
              >
                <i
                  className="icon icon--material "
                  data-qa-file="Icon"
                  data-qa-node="i"
                >
                  keyboard_arrow_down
                </i>
                Track number
              </div>
              <div
                className="dropdown-field__options__item "
                data-qa-file="DropdownField"
                data-qa-node="div"
                onClick={[Function]}
              >
                Name
              </div>
            </div>
          </div>
        </div>
      </div>
    </h4>
  </section>
  <section
    className="wiki"
    data-qa-file="Album"
    data-qa-node="section"
  >
    <h4
      className="wiki__title"
      data-qa-file="Album"
      data-qa-node="h4"
    >
      About
    </h4>
    <div
      className="wiki__text"
      data-qa-file="Album"
      data-qa-node="div"
    >
      <p
        dangerouslySetInnerHTML={
          {
            "__html": "Wiki text",
          }
        }
        data-qa-file="Album"
        data-qa-node="p"
      />
      <br
        data-qa-file="Album"
        data-qa-node="br"
      />
      <div
        className="mid_grey-text"
        data-qa-file="Album"
        data-qa-node="div"
      >
        Published: 2021-01-01
      </div>
    </div>
  </section>
</div>
`;
